File: 0015-3sum/solution.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        negative = []
        positive = []
        zeros = 0
        
        for num in nums:
            if num > 0:
                positive.append(num)
            elif num == 0:
                zeros += 1
            else:
                negative.append(num)
        
        result = []
        encounteredTuples = set()
        
        negativeSet = set(negative)
        positiveSet = set(positive)
        
        if zeros >= 3:
            encounteredTuples.add((0,0,0))
        
        # 2 positive 1 negative
        for p1, p2 in itertools.combinations(positive, 2):
            if -(p1 + p2) in negativeSet:
                item = tuple(sorted([p1, p2, -(p1+p2)]))
                encounteredTuples.add(item)
        
        # 2 negative 1 positive
        for n1, n2 in itertools.combinations(negative, 2):
            if -(n1 + n2) in positiveSet:
                item = tuple(sorted([n1, n2, -(n1+n2)]))
                encounteredTuples.add(item)
        
        # 1 positive 1 negative 1 zero
        if zeros > 0:
            for p in positive:
                if -p in negativeSet:
                    item = tuple([-p, 0, p])
                    encounteredTuples.add(item)

        return [list(item) for item in encounteredTuples]

File: 0015-3sum/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;

        sort(nums.begin(), nums.end());

        const int n = nums.size();

        for (int i = 0; i < n; i++) {
            if (i > 0 && nums[i - 1] == nums[i]) continue;
            
            int left = i + 1;
            int right = n - 1;

            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];

                if (sum > 0) {
                    right--;
                } else if (sum < 0) {
                    left++;
                } else {
                    ans.push_back({nums[i], nums[left], nums[right]});
                    left++;
                    right--;
                    while (left < right && nums[left - 1] == nums[left]) {
                        left++;
                    }
                }
            }
        }

        return ans;

    }
};
